### PR TITLE
Purchases: Remove PurchaseCardDetails inheritance

### PR DESCRIPTION
**Pull request by sirreal**
Remove `PurchaseCardDetails`, which subclasses `React.Component`, but is not a renderable `Component` instead is intended for inheritance.

[React docs discourage inheritance](https://reactjs.org/docs/composition-vs-inheritance.html#so-what-about-inheritance) and it is not used as a convention in Calypso components. Indeed, the use of this class obscured prop usage and only saved a bit of code duplication.

This PR takes the simple approach of duplicating a few methods and removing the intermediate class.

An alternative would be to add a specific component containing `CreditCardForm`.

## Testing
1. [(/me/purchases/)](https://calypso.live/me/purchases?branch=remove/purchasecarddetails-superclass&debug)
1. `localStorage.debug = "calypso:analytics:tracks"`
1. Pick a purchase (purchased with credits is a good idea).
1. Add a credit card. Does the form work as expected?
1. Edit the credit card. Does it continue to work well?
1. Was the `calypso_purchases_credit_card_form_submit` event fired on form submission?

Prep for #25219

### CODE REVIEW

File: client/me/purchases/payment/add-card-details/index.jsx [hunk 1]
```diff
@@ -3,42 +3,76 @@
...
 import PropTypes from 'prop-types';
-import React from 'react';
+import React, { Component, Fragment } from 'react';
+import { connect } from 'react-redux';
 
 /**
  * Internal Dependencies
  */
-import { clearPurchases } from 'state/purchases/actions';
+import analytics from 'lib/analytics';
```
Comment (sirreal) [inline client/me/purchases/payment/add-card-details/index.jsx]: Would be good to get off `lib/analytics`

Comment (sirreal) [reply client/me/purchases/payment/add-card-details/index.jsx]: ✅ 675be8a

File: client/me/purchases/payment/add-card-details/index.jsx [hunk 2]
```diff
@@ -3,42 +3,75 @@
...
+	}
+
+	isDataValid( props = this.props ) {
+		const { purchase, selectedSite } = props;
+
+		return purchase && selectedSite;
+	}
+
+	recordFormSubmitEvent = () =>
+		void this.props.recordTracksEvent( 'calypso_purchases_credit_card_form_submit', {
```
Comment (ockham) [inline client/me/purchases/payment/add-card-details/index.jsx]: Hmm, what's the `void` doing here?

File: client/me/purchases/payment/edit-card-details/index.jsx [hunk 3]
```diff
@@ -38,11 +40,42 @@ class EditCardDetails extends PurchaseCardDetails {
...
+	}
+
+	isDataValid( props = this.props ) {
+		const { purchase, selectedSite } = props;
+
+		return purchase && selectedSite;
+	}
+
+	recordFormSubmitEvent = () =>
+		void this.props.recordTracksEvent( 'calypso_purchases_credit_card_form_submit', {
```
Comment (ockham) [inline client/me/purchases/payment/edit-card-details/index.jsx]: https://github.com/Automattic/wp-calypso/pull/25442/files#r195436963

**Comment by sirreal** [comment]
> Are you planning anything for getting rid of the duplication that we create here?

Not immediately. This is a nudge along in a simpler, clearer direction. I acknowledge that there is more work to be done 🙂

**Comment by dmsnell** [comment]
Not reverted on account of seeing a false alarm. The revert PR mentioned ^^^ (#25578) was prepped in order to be ready but not needed.

**Pull request merged**